src/main/java/net/minestom/server/inventory/AbstractInventory.java

```diff
@@ -0,0 +1,366 @@
+            itemChangesMap.forEach(this::safeItemInsert);
+            return true;
+        } else {
+            // Inventory cannot accept the item fully
+            return false;
+        }
+    }
+
+    /**
+     * Adds {@link ItemStack}s to the inventory and sends relevant updates to the viewer(s).
+     * <p>
+     * Even items cannot be fully added, the amount of {@code itemStack}s will be updated.
+     *
+     * @param itemStacks items to add
+     * @return list of itemstacks that could not be successfully fully added, empty list otherwise
+     */
+    public List<ItemStack> addItemStacks(@NotNull List<ItemStack> itemStacks) {
+        List<ItemStack> result = new ArrayList<>();
+        itemStacks.forEach(itemStack -> {
+            if (!addItemStack(itemStack)) {
+                result.add(itemStack);
+            }
+        });
+        return result;
+    }
+
+    /**
+     * Checks whether {@link ItemStack} can be fully added to the inventory.
+     *
+     * @param itemStack the item to be checked
+     * @return true if the item can be fully added to the inventory, false otherwise
+     */
+    public boolean canAddItemStack(@NotNull ItemStack itemStack) {
+        final StackingRule stackingRule = itemStack.getStackingRule();
+        int amountLeft = stackingRule.getAmount(itemStack);
+        for (int i = 0; i < getInnerSize(); i++) {
+            ItemStack inventoryItem = getItemStack(i);
+            if (stackingRule.canBeStacked(itemStack, inventoryItem)) {
+                final int itemAmount = stackingRule.getAmount(inventoryItem);
+                if (itemAmount == stackingRule.getMaxSize())
+                    continue;
+                if (!stackingRule.canApply(itemStack, amountLeft + itemAmount)) {
+                    // Slot cannot accept the whole item, reduce amount to 'itemStack'
+                    amountLeft -= stackingRule.getMaxSize() - itemAmount;
+                } else {
+                    return true;
+                }
+            } else if (inventoryItem.isAir()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Checks whether {@link ItemStack}s can be fully added to the inventory.
+     *
+     * @param itemStacks items to be checked
+     * @return true if all the items can be fully added to the inventory, false otherwise
+     */
+    public boolean canAddItemStacks(@NotNull List<ItemStack> itemStacks) {
+        return itemStacks.stream().allMatch(this::canAddItemStack);
+    }
+
+    /**
+     * Takes an {@link ItemStack} from the inventory and sends relevant update to the viewer(s).
+     * <p>
+     * Even the item cannot be fully taken, the amount of {@code itemStack} will be updated.
+     *
+     * @param itemStack the item to take
+     * @return true if the item has been successfully fully taken, false otherwise
+     */
+    public boolean takeItemStack(@NotNull ItemStack itemStack) {
```

Comment:
Same with this, it should return an `ItemStack` containing what couldn't be taken.